File: moling-server/app/router/notification.py
"""墨灵 (Moling) — 通知 API 路由。

实现列出通知、标记已读、删除通知等端点。
"""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_db, get_current_user
from app.service.notification_service import notification_service
from app.schemas.notification import NotificationResp

router = APIRouter()


@router.get("", response_model=dict)
async def list_notifications(
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    is_read: bool = Query(None, description="已读状态筛选"),
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
) -> dict:
    """获取当前用户的通知列表。"""
    result = await notification_service.list_notifications(
        db,
        current_user.id,
        page=page,
        page_size=page_size,
        is_read=is_read,
    )
    return result


@router.get("/unread-count", response_model=dict)
async def get_unread_count(
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
) -> dict:
    """获取未读通知数量。"""
    result = await notification_service.get_unread_count(
        db,
        current_user.id,
    )
    return result


@router.post("/{notification_id}/read", response_model=NotificationResp)
async def mark_as_read(
    notification_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
) -> NotificationResp:
    """标记通知为已读。"""
    result = await notification_service.mark_as_read(
        db,
        current_user.id,
        notification_id,
    )
    return result


@router.post("/read-all", response_model=dict)
async def mark_all_as_read(
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
) -> dict:
    """标记所有通知为已读。"""
    result = await notification_service.mark_all_as_read(
        db,
        current_user.id,
    )
    return result


@router.delete("/{notification_id}", status_code=204)
async def delete_notification(
    notification_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
) -> None:
    """删除通知。"""
    await notification_service.delete_notification(
        db,
        current_user.id,
        notification_id,
    )

File: moling-server-rs/crates/moling-api/src/routes/notification.rs
//! Notification routes — list, unread count, read, read-all, delete.

use axum::{extract::{Path, Query, State}, routing::{delete, get, post}, Json, Router};
use moling_core::error::{AppError, AppResult};
use moling_db::dao::notification_dao::NotificationDao;
use moling_auth::CurrentUser;
use crate::state::AppState;
use crate::types::*;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", get(list_notifications))
        .route("/unread-count", get(unread_count))
        .route("/{id}/read", post(mark_as_read))
        .route("/read-all", post(mark_all_read))
        .route("/{id}", delete(delete_notification))
}

async fn list_notifications(
    State(state): State<AppState>,
    user: CurrentUser,
    Query(q): Query<NotificationQuery>,
) -> AppResult<Json<serde_json::Value>> {
    let dao = NotificationDao;
    let skip = (q.page as u64 - 1) * q.page_size as u64;
    let items = dao.list_by_user(&state.db, &user.user_id.to_string(), skip, q.page_size as u64, q.is_read).await?;
    let total = dao.count_by_user(&state.db, &user.user_id.to_string(), q.is_read).await?;
    Ok(Json(serde_json::json!({ "items": items, "total": total, "page": q.page, "page_size": q.page_size })))
}

async fn unread_count(
    State(state): State<AppState>,
    user: CurrentUser,
) -> AppResult<Json<serde_json::Value>> {
    let dao = NotificationDao;
    let count = dao.count_by_user(&state.db, &user.user_id.to_string(), Some(false)).await?;
    Ok(Json(serde_json::json!({ "unread_count": count })))
}

async fn mark_as_read(
    State(state): State<AppState>,
    user: CurrentUser,
    Path(id): Path<i32>,
) -> AppResult<Json<serde_json::Value>> {
    let dao = NotificationDao;
    dao.mark_as_read(&state.db, id, &user.user_id.to_string()).await?.ok_or_else(|| AppError::not_found("通知不存在".to_owned()))?;
    Ok(Json(serde_json::json!({ "message": "Marked as read" })))
}

async fn mark_all_read(
    State(state): State<AppState>,
    user: CurrentUser,
) -> AppResult<Json<serde_json::Value>> {
    let dao = NotificationDao;
    let count = dao.mark_all_as_read(&state.db, &user.user_id.to_string()).await?;
    Ok(Json(serde_json::json!({ "updated": count })))
}

async fn delete_notification(
    State(state): State<AppState>,
    user: CurrentUser,
    Path(id): Path<i32>,
) -> AppResult<axum::http::StatusCode> {
    let dao = NotificationDao;
    dao.delete_by_user(&state.db, id, &user.user_id.to_string()).await?;
    Ok(axum::http::StatusCode::NO_CONTENT)
}
